src/DotNetCore.CAP.RabbitMQ/ITransport.RabbitMQ.cs

```diff
@@ -2,74 +2,116 @@
-                    Description = ex.Message
-                };
-
-                return Task.FromResult(OperateResult.Failed(wrapperEx, errors));
-            }
-            finally
-            {
-                if (channel != null)
+                    var activityName = message.GetName();
+    
+                    using (var activity = ActivitySource.StartActivity(activityName, ActivityKind.Producer))
+                    {
+                        channel = _connectionChannelPool.Rent();
+    
+                        channel.ConfirmSelect();
+    
+                        var props = channel.CreateBasicProperties();
+                        props.DeliveryMode = 2;
+                        props.Headers = message.Headers.ToDictionary(x => x.Key, x => (object) x.Value);
+                        
+                        ActivityContext contextToInject = default;
+                        if (activity != null)
+                        {
+                            contextToInject = activity.Context;
+                        }
+                        else if (Activity.Current != null)
+                        {
+                            contextToInject = Activity.Current.Context;
+                        }
+    
+                        Propagator.Inject(new PropagationContext(contextToInject, Baggage.Current), props, InjectTraceContextIntoBasicProperties);
+    
+                        AddMessagingTags(activity, message);
+                        
+                        channel.WaitForConfirmsOrDie(TimeSpan.FromSeconds(5));
+    
+                        channel.ExchangeDeclare(_exchange, RabbitMQOptions.ExchangeType, true);
+    
+                        channel.BasicPublish(_exchange, message.GetName(), props, message.Body);
+    
+                        _logger.LogDebug($"RabbitMQ topic message [{message.GetName()}] has been published.");
+    
+                        return Task.FromResult(OperateResult.Success);
+                    }
+                }
+                catch (System.Exception ex)
+                {
+                    var wrapperEx = new PublisherSentFailedException(ex.Message, ex);
+                    var errors = new OperateError
+                    {
+                        Code = ex.HResult.ToString(),
+                        Description = ex.Message
+                    };
+    
+                    return Task.FromResult(OperateResult.Failed(wrapperEx, errors));
+                }
+                finally
                 {
-                    _connectionChannelPool.Return(channel);
+                    if (channel != null)
+                    {
+                        _connectionChannelPool.Return(channel);
+                    }
                 }
             }
+            
+            private void AddMessagingTags(Activity activity, TransportMessage message)
+            {
+                activity?.SetTag("message", Encoding.UTF8.GetString(message.Body.ToArray()));
```

Comment:
I see where you coming from here, but should the message really be logged?! Imagine a scenario where a message contains PII and you cannot log it due to GDPR concerns.

Reply:
@mviegas Yes, It's true, I removed this line and add correlation_id and message_id instead of it.